src/modules/Popup/Popup.js

```diff
@@ -0,0 +1,226 @@
+
+    if (_.includes(positions, 'right')) {
+      style.right = Math.round(clientWidth - (this.coords.right + pageXOffset))
+      style.left = 'auto'
+    } else if (_.includes(positions, 'left')) {
+      style.left = Math.round(this.coords.left + pageXOffset)
+      style.right = 'auto'
+    } else {  // if not left nor right, we are horizontally centering the element
+      const xOffset = (this.coords.width - this.popupCoords.width) / 2
+      style.left = Math.round(this.coords.left + xOffset + pageXOffset)
+      style.right = 'auto'
+    }
+
+    if (_.includes(positions, 'top')) {
+      style.bottom = Math.round((clientHeight - pageYOffset) - this.coords.top)
+      style.top = 'auto'
+    } else if (_.includes(positions, 'bottom')) {
+      style.top = Math.round(this.coords.bottom + pageYOffset)
+      style.bottom = 'auto'
+    } else {  // if not top nor bottom, we are vertically centering the element
+      const yOffset = (this.coords.height + this.popupCoords.height) / 2
+      style.top = Math.round(this.coords.bottom + pageYOffset - yOffset)
+      style.bottom = 'auto'
+
+      const xOffset = this.popupCoords.width + 8
+      if (_.includes(positions, 'right')) {
+        style.right -= xOffset
+      } else {
+        style.left -= xOffset
+      }
+    }
+
+    return style
+  }
+
+  // returns an adjusted position according to a given style
+  // so the popover always appear fully visible
+  correctPositionForViewPort(style) {
+    const { clientHeight } = document.documentElement
+    const { pageYOffset, pageXOffset } = window
+
+    if (_.isNumber(style.right)) {
+      if (style.right + pageXOffset < 0) {
+        return 'left center'
+      }
+    } else if (style.left - pageXOffset < 0) {
+      return 'right center'
+    }
+
+    if (_.isNumber(style.bottom)) {
+      if (style.bottom + this.popupCoords.height > clientHeight - pageYOffset) {
+        return 'bottom center'
+      }
+    } else if (style.top + this.popupCoords.height > pageYOffset + clientHeight) {
+      return 'top center'
+    }
+  }
+
+  getPopupStyle() {
+    if (!this.coords || !this.popupCoords) return
+    let position = this.props.position
+    let style = this.computePopupStyle(position)
+
+    // Lets detect if the popup is out of the viewport and adjust
+    // the position accordingly
+    const correctedPosition = this.correctPositionForViewPort(style)
+    if (correctedPosition) {
+      style = this.computePopupStyle(correctedPosition)
+      // Set the position so the carret will appear at the right place
+      position = correctedPosition
+    }
+
+    // Append 'px' to every numerical values in the style
+    style = _.mapValues(style, value => _.isNumber(value) ? value + 'px' : value)
+
+    this.setState({ style, position })
+  }
+
+  onOpen = (event) => {
+    this.coords = event.currentTarget.getBoundingClientRect()
+  }
+
+  componentWillMount() {
+    const portalProps = {}
+
+    if (this.props.on === 'click') {
```

Comment:
I would move this out of `componentWillMount` and out of state. This is only called on initial mount, so if the `on` props change then this will be stale. I'd do something like:
```js
// At the top of the file
const triggerActionToPortalProps = {
  click: {
    openOnTriggerClick: true,
    closeOnDocumentClick: true,
  },
  hover: {
    openOnTriggerMouseOver: true,
    closeOnTriggerMouseLeave: true,
  }
  focus: {
    openOnTriggerFocus: true,
    closeOnTriggerBlur: true,
  }
}

// Within render:
const { on } = this.props

<Portal trigger={trigger} {...onToPortalProps[on]}>
```